### ISSUE TITLE
 The model always outputs the same label.

### ISSUE THREAD

**Issue by RoyZhDec** [open-mmlab/mmsegmentation#1134]
I use cityscapes dataset to train a 2_class segmentor, and I edited the self.label_map according to my needs, shown as below.
![image](https://user-images.githubusercontent.com/20221570/145951275-f853c6ec-bd3b-4664-b16c-ae5af3a83cea.png)

The loss_ce and acc_seg values ​​during the entire training process were normal, and the visualization results were normal at first, but as the loss decreased, it became more and more strange. 
![CFB50005-08E9-4230-A4F7-FD622806DD02](https://user-images.githubusercontent.com/20221570/145951097-5c44e3c9-5047-470b-a508-101fee8fc0ad.png)

**Comment by MengzhangLI** [comment]
Hi, could you describe the strange things in figure?

**Comment by RoyZhDec** [comment]
> Hi, could you describe the strange things in figure?

the color below represents road
![image](https://user-images.githubusercontent.com/20221570/145991590-095b4668-a84b-4568-b33f-0e8673f53873.png)
the color below represents background
![image](https://user-images.githubusercontent.com/20221570/145991633-cb8949b9-2066-412d-86d5-0fc6db5ea919.png)

So in the epoch20 image and epoch40 image, their mask is well explained and it makes sense. But after epoch40, the mask always tends to predict as background, even if loss is falling normally, acc is rising normally, up to 88%.

**Comment by maisonhai3** [comment]
@RoyZhDec Please give us a look at your configs.

**Comment by RoyZhDec** [comment]
> @RoyZhDec Please give us a look at your configs.

norm_cfg = dict(type='BN', requires_grad=True)
has_bn = True
model = dict(
    type='EncoderDecoder2A4',
    pretrained='/home/zqs/DLE/pretrained_model/tiny-yolo/tiny-yolo.pth.tar',
    backbone=dict(
        type='DLETinyYolo',
        num_stages=5,
        basicblocks=(('DLEConvBNRelu', ), ('DLEConvBNRelu', 'DLEConvBNRelu',
                                           'DLEConvBNRelu'),
                     ('DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu', 'DLEConvBNRelu'),
                     ('DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu'),
                     ('DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu', 'DLEConvBNRelu', 'DLEConvBNRelu',
                      'DLEConvBNRelu')),
        inplanes=((3, ), (32, 64, 32), (64, 128, 64, 128, 64),
                  (128, 256, 128, 256, 128, 256,
                   128), (256, 512, 256, 512, 256, 512, 256, 512, 256, 512,
                          256, 512, 256)),
        planes=((32, ), (64, 32, 64), (128, 64, 128, 64, 128),
                (256, 128, 256, 128, 256, 128,
                 256), (512, 256, 512, 256, 512, 256, 512, 256, 512, 256, 512,
                        256, 512)),
        kernel_sizes=((3, ), (3, 1, 3), (3, 1, 3, 1, 3), (3, 1, 3, 1, 3, 1, 3),
                      (3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3)),
        strides=((2, ), (2, 1, 1), (2, 1, 1, 1, 1), (2, 1, 1, 1, 1, 1, 1),
                 (2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)),
        paddings=((1, ), (1, 0, 1), (1, 0, 1, 0, 1), (1, 0, 1, 0, 1, 0, 1),
                  (1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1)),
        dilations=((1, ), (1, 1, 1), (1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1),
                   (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)),
        groups=((1, ), (1, 1, 1), (2, 1, 2, 1, 2), (4, 1, 4, 1, 4, 1, 4),
                (8, 1, 8, 1, 8, 1, 8, 1, 8, 1, 8, 1, 8)),
        out_indices=(2, 4),
        frozen_stages=-1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        has_bn=True),
    decode_head=dict(
        type='TinyYolo2A4Head',
        inplanes=(512, 128, 64, 32, 128),
        planes=(128, 64, 32, 32, 64),
        kernel_sizes=(1, 3, 1, 1, 1),
        strides=(1, 1, 1, 1, 1),
        paddings=(0, 1, 0, 0, 0),
        dilations=(1, 1, 1, 1, 1),
        upsample_mode=('bilinear', 'bilinear', 'bilinear'),
        dropout_ratio=0,
        num_classes=2,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0)),
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))
dataset_type = 'CityscapesDataset_customed'
data_root = '/home/zqs/datasets/cityscapes/'
img_norm_cfg = dict(mean=[128, 128, 128], std=[128, 128, 128], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=(1280, 736), keep_ratio=False),
    dict(type='RandomFlip', prob=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(
        type='Normalize',
        mean=[128, 128, 128],
        std=[128, 128, 128],
        to_rgb=True),
    dict(type='Pad', size_divisor=32, pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1280, 736),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=False),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[128, 128, 128],
                std=[128, 128, 128],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=8,
    workers_per_gpu=2,
    train=dict(
        type='CityscapesDataset_customed',
        data_root='/home/zqs/datasets/cityscapes/',
        img_dir='leftImg8bit/train',
        ann_dir='gtFine/train',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations'),
            dict(type='Resize', img_scale=(1280, 736), keep_ratio=False),
            dict(type='RandomFlip', prob=0.5),
            dict(type='PhotoMetricDistortion'),
            dict(
                type='Normalize',
                mean=[128, 128, 128],
                std=[128, 128, 128],
                to_rgb=True),
            dict(type='Pad', size_divisor=32, pad_val=0, seg_pad_val=255),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_semantic_seg'])
        ]),
    val=dict(
        type='CityscapesDataset_customed',
        data_root='/home/zqs/datasets/cityscapes/',
        img_dir='leftImg8bit/val',
        ann_dir='gtFine/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1280, 736),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=False),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[128, 128, 128],
                        std=[128, 128, 128],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]),
    test=dict(
        type='CityscapesDataset_customed',
        data_root='/home/zqs/datasets/cityscapes/',
        img_dir='leftImg8bit/val',
        ann_dir='gtFine/val',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1280, 736),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=False),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[128, 128, 128],
                        std=[128, 128, 128],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]))
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
lr_config = dict(
    policy='poly',
    warmup='linear',
    warmup_iters=1000,
    warmup_ratio=0.001,
    power=0.9)
runner = dict(type='EpochBasedRunner', max_epochs=500)
checkpoint_config = dict(interval=10)
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=False),
        dict(type='TensorboardLoggerHook')
    ])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
work_dir = './class2_no_drop/'
gpu_ids = range(0, 1)

**Comment by RoyZhDec** [comment]
> I think you could do somethings:
> 
> (1) Do validation to testify the prediction result, to check whether model itself is correct. If model is correct, then it is caused by visulization.
> 
> (2) Try to use [customized dataset](https://github.com/open-mmlab/mmsegmentation/blob/master/docs_zh-CN/tutorials/customize_datasets.md) becuase I think it probably caused by citycapes 19 num classes or your config file.
> 
> Best,


I used the mmseg official cityscapes dataset code and used my own model to train all 19 types of labels. The visualization results and eval results are very poor(shown as below). Perhaps this means that my model itself has a problem? But the loss_ce and acc_seg indicators both show that the model is well trained. Why?

![frankfurt_000000_000294_leftImg8bit](https://user-images.githubusercontent.com/20221570/146344788-e2e3eb1f-1292-4c5a-9757-9bb87e7b8be5.png)

+---------------+-------+-------+
|     Class     |  IoU  |  Acc  |
+---------------+-------+-------+
|      road     |  4.22 |  4.33 |
|    sidewalk   |  0.0  |  0.0  |
|    building   |  0.33 |  0.33 |
|      wall     |  1.04 | 32.55 |
|     fence     |  0.18 |  0.18 |
|      pole     | 15.23 | 30.05 |
| traffic light |  4.32 |  5.89 |
|  traffic sign | 10.73 | 14.87 |
|   vegetation  | 27.96 |  93.0 |
|    terrain    |  2.87 | 46.72 |
|      sky      |  24.1 | 27.28 |
|     person    |  0.23 |  0.23 |
|     rider     |  0.01 |  0.01 |
|      car      | 13.23 | 14.81 |
|     truck     |  0.01 |  0.01 |
|      bus      |  0.03 |  0.03 |
|     train     |  0.0  |  0.0  |
|   motorcycle  |  0.0  |  0.0  |
|    bicycle    |  0.01 |  0.01 |
+---------------+-------+-------+
Summary:

+-------+------+-------+
|  aAcc | mIoU |  mAcc |
+-------+------+-------+
| 20.87 | 5.5  | 14.23 |
+-------+------+-------+

**Comment by MengzhangLI** [comment]
I think there is something wrong in https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/datasets/cityscapes.py but right now I can not give clear answer. 

Could you discard provided cityscapes dataset and try to use customized dataset preparation? Given `num_classes` in cityscapes is 19 but yours are 2.

**Comment by RoyZhDec** [comment]
> I think there is something wrong in https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/datasets/cityscapes.py but right now I can not give clear answer.
> 
> Could you discard provided cityscapes dataset and try to use customized dataset preparation? Given `num_classes` in cityscapes is 19 but yours are 2.

I find the problem. My training is all right, because I have shown the inference result during the training, and the training-time inference pics just show good segmentation result(shown as pic1 below, black means road and white means background). So I think it's something wrong with my test.py. Then I print my model parameter in test.py, and I find one of the BN layer is invalid(shown as pic2 below). I think it's why all my test result pixel is label 1, which leads to a all white segmentation picture . However, I still can't figure out why the BN layer is ok during training,  then become invalid in testing. Any help?THANKS!

pic1:
![0 0005204677581787109](https://user-images.githubusercontent.com/20221570/146501638-03726d38-66a1-447c-b937-411ceab80825.jpg)

pic2:
![image](https://user-images.githubusercontent.com/20221570/146501591-35128c0d-aa00-4403-a4da-8dea36166bfa.png)

**Comment by MengzhangLI** [comment]
Very happy to see you have solved the problem. ;)

**Comment by Franko9999** [comment]
Hello,

So how did you solve the problem. I have the same one for a two class custom dataset.

Best regards,
Roberts

**Comment by Franko9999** [comment]
Hello,

I have the same problem, my custom dataset have only two classes (simple baloon dataset). The inference always predicts only background class. I have tried with both reduce_zero_label=True and reduce_zero_label_False. It would be nice if someone could help me in this.

the sample prediction (only darkens the image as background class):

![result](https://user-images.githubusercontent.com/47304445/185583307-d0e385d9-9b09-4253-9749-cb87641936d8.jpg)


Best regards,
Roberts
